websockets/protocol/caret.go

```diff
@@ -0,0 +1,81 @@
+package protocol
+
+type Caret struct {
+	Start int `json:"caret_start"`
+	End   int `json:"caret_end"`
+}
+
+func processAdd(
+	receiverCaret Caret,
+	senderCaret Caret,
+	caretStartDelta int,
+	caretEndDelta int,
+	docDelta int,
+) Caret {
+	if senderCaret.End < receiverCaret.End && caretEndDelta != 0 {
+		receiverCaret.End += docDelta
+		if senderCaret.End <= receiverCaret.Start {
+			receiverCaret.Start += docDelta
+		}
+	}
+	return receiverCaret
+}
+
+func ShiftCaret(
+	receiverCaret Caret,
+	senderCaret Caret,
+	caretStartDelta int,
+	caretEndDelta int,
+	docDelta int,
+) Caret {
+	var rangeStart, rangeEnd int
+	if senderCaret.Start != senderCaret.End {
+		rangeStart = senderCaret.Start
+		rangeEnd = senderCaret.End
+	} else if docDelta > 0 {
+		rangeStart = senderCaret.Start
+		rangeEnd = senderCaret.Start + caretStartDelta
+	} else if caretStartDelta == 0 {
+		rangeStart = senderCaret.Start
+		rangeEnd = senderCaret.Start - docDelta
+	} else {
+		rangeStart = senderCaret.Start + caretStartDelta
+		rangeEnd = senderCaret.Start
+	}
+
+	if docDelta > 0 && senderCaret.Start == senderCaret.End {
+		return processAdd(receiverCaret, senderCaret, caretStartDelta, caretEndDelta, docDelta)
+	}
+
+	delta := rangeStart - rangeEnd
+
+	if rangeEnd < receiverCaret.End {
+		receiverCaret.End += delta
+		if rangeEnd <= receiverCaret.Start {
+			receiverCaret.Start += delta
+		} else {
+			if rangeStart < receiverCaret.Start {
+				receiverCaret.Start = rangeStart
+			}
+		}
```

Comment:
This can be changed to an `else if`.
```suggestion
		} else if rangeStart < receiverCaret.Start {
			receiverCaret.Start = rangeStart
		}
```